stdlib/public/core/StringNormalization.swift

```diff
@@ -102,3 +102,253 @@ internal func _tryNormalize(
   }
   return numericCast(count)
 }
+
+internal enum NormalizationResult {
+  case success(SuccessResult)
+  case bufferTooSmall(BufferResizeRequest) // The size needed to normalize the rest of the string
+  
+  struct SuccessResult {
+    var amountFilled: Int
+    var nextReadPosition: String.Index
+  }
+  struct BufferResizeRequest {
+    var outputBufferSize: Int
+    var icuInputBufferSize: Int
+    var icuOutputBufferSize: Int
+  }
+  
+  static func bufferTooSmall(count: Int) -> NormalizationResult {
+    let outputBufferSize = 
+      count * _Normalization._maxNFCExpansionFactor * _Normalization._maxUTF16toUTF8ExpansionFactor
+    let icuInputBufferSize = count
+    let icuOutputBufferSize = count * _Normalization._maxNFCExpansionFactor
+    let resizeRequest = BufferResizeRequest(
+      outputBufferSize: outputBufferSize, 
+      icuInputBufferSize: icuInputBufferSize,
+      icuOutputBufferSize: icuOutputBufferSize
+    )
+    return .bufferTooSmall(resizeRequest)
+  }
+  
+  static func success(
+    amountFilled filled: Int, nextReadPosition index: String.Index
+  ) -> NormalizationResult {
+    let successResult = SuccessResult(amountFilled: filled, nextReadPosition: index)
+    return .success(successResult)
+  }
+}
+
+internal func fastFill(
+  _ sourceBuffer: UnsafeBufferPointer<UInt8>,
+  _ outputBuffer: UnsafeMutableBufferPointer<UInt8>
+) -> (Int, Int) {
+  // Quick check if a scalar is NFC and a segment starter
+  @inline(__always) func isNFCStarter(_ scalar: Unicode.Scalar) -> Bool {
+    // Fast-path: All scalars up through U+02FF are NFC and have boundaries
+    // before them
+    if scalar.value < 0x300 { return true }
+
+    // Otherwise, consult the properties
+    return scalar._hasNormalizationBoundaryBefore && scalar._isNFCQCYes
+  }
+  
+  var outputBufferThreshold: Int {
+    return outputBuffer.count - 4
+  }
+
+  // TODO: Additional fast-path: All CCC-ascending NFC_QC segments are NFC
+  // TODO: Just freakin do normalization and don't bother with ICU
+  var outputCount = 0
+  let outputEnd = outputBufferThreshold
+  var inputCount = 0
+  let inputEnd = sourceBuffer.count
+  while inputCount < inputEnd && outputCount < outputEnd {
+    // TODO: Slightly faster code-unit scan for latiny (<0xCC)
+
+    // Check scalar-based fast-paths
+    let (scalar, len) = _decodeScalar(sourceBuffer, startingAt: inputCount)
+    _internalInvariant(inputCount &+ len <= inputEnd)
+
+    if _slowPath(
+         !sourceBuffer.hasNormalizationBoundary(before: inputCount &+ len)
+      || !isNFCStarter(scalar)
+    ) {
+      break
+    }
+    inputCount &+= len
+
+    for cu in UTF8.encode(scalar)._unsafelyUnwrappedUnchecked {
+      outputBuffer[outputCount] = cu
+      outputCount &+= 1
+    }
+
+    _internalInvariant(inputCount == outputCount,
+      "non-normalizing UTF-8 fast path should be 1-to-1 in code units")
+  }
+  return (inputCount, outputCount)
+}
+
+internal func transcodeToUTF16(
```

Comment:
This function is specifically transcoding a segment. Please adjust the name to reflect that and add a comment, otherwise it looks like a general purpose transcoding function.